### PR TITLE
improvement(setup): Parallelize setup

### PR DESCRIPTION
**Pull request by soyacz**
In order to speed up test setup, made setup of db cluster, loaders,
monitors and oracle cluster to configure in parallel.
Verification and Argus related work was put into one-off background
thread.

Removed dependency to db_cluster for loaders as was not used in code (if needed we can add it in different place).

refs: https://github.com/scylladb/scylla-cluster-tests/issues/7242

### Testing

- [x] - [gemini test](https://argus.scylladb.com/test/5b50c349-c40c-440c-b727-f69c2e00acb4/runs?additionalRuns[]=1e020da2-3a83-4812-920d-3b51207c2693) - provision passed, errors due lost connectivity to loader after some time. More details in argus.
- [x] - provision test (jenkins pipeline, all backends passed including k8s functional tests)
- [x] - [k8s multitenant](https://argus.scylladb.com/test/80b4b4e8-be74-4439-98a8-889102638f31/runs?additionalRuns[]=e374e4cd-39a6-4822-bdee-08a7824c2e7c) - mutation write error during short period of time only. 
- [X] - [artifact debian](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/artifacts-debian11-test/3/)
- [X] - [artifact ami](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/artifacts-ami-test/3/)
- [x] - [ms ldap](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/longevity-100gb-4h-ldap-authorization-authenticator-ms-ad-test/3/)

### PR pre-checks (self review)

- [ ] I added the relevant `backport` labels
- [ ] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

### CODE REVIEW

File: sdcm/tester.py
Comment (vponomaryov) [inline sdcm/tester.py:813]: Why is it changed?

File: sdcm/tester.py
Comment (vponomaryov) [inline sdcm/tester.py:910]: This looks racy.
Probably will fail when LDAP setup is enabled.

File: sdcm/tester.py
Comment (vponomaryov) [inline sdcm/tester.py:912]: Same here.

File: sdcm/tester.py
Comment (vponomaryov) [inline sdcm/tester.py:961]: Error verification?

Comment (fruch) [reply sdcm/tester.py:961]: *Let's check if _validate_raft is needed to be wait for,
* lets try moving to using  `ThreadPoolExecutor`

Comment (soyacz) [reply sdcm/tester.py]: ok, I decided for now to move this under ThreadPoolExecutor as running it in background thread and not waiting or result requires analysis. This will be done in future work for speeding up setup.

File: sdcm/tester.py
Comment (soyacz) [reply sdcm/tester.py]: I moved it to separate block after setup db/loaders/monitors in sequential way - should not impact setup time.

File: sdcm/tester.py
Comment (soyacz) [reply sdcm/tester.py]: mistake

**Review by vponomaryov — DISMISSED** [review]
LGTM if it works.

**Comment by fruch** [comment]
> LGTM if it works.

@soyacz let's have a more detailed plan on what needs to be covered.

I think we should try at least one type of testing, or wait for 6.0 branching before merging this.

**Comment by soyacz** [comment]
> > LGTM if it works.
> 
> @soyacz let's have a more detailed plan on what needs to be covered.
> 
> I think we should try at least one type of testing, or wait for 6.0 branching before merging this.

I was testing it with provision test and analyzing logs if everything goes ok. I also executed Gemini test (as also oracle got parallelized). I was planning to run k8s multitenant case as it is not covered by provision tests triggered in by our SCT pipeline.

Anything else comes up to your mind?

**Comment by fruch** [comment]
On Wed, May 8, 2024, 08:06 Lukasz Sojka ***@***.***> wrote:

> LGTM if it works.
>
> @soyacz <https://github.com/soyacz> let's have a more detailed plan on
> what needs to be covered.
>
> I think we should try at least one type of testing, or wait for 6.0
> branching before merging this.
>
> I was testing it with provision test and analyzing logs if everything goes
> ok. I also executed Gemini test (as also oracle got parallelized). I was
> planning to run k8s multitenant case as it is not covered by provision
> tests triggered in by our SCT pipeline.
>
> Anything else comes up to your mind?
>

Checking artifact test wasn't broken in the process

File: sdcm/tester.py
Comment (fruch) [inline sdcm/tester.py:942]: This is not related to raft

And it starts a thread of its own, no point putting it into a thread

But anyhow the title of the function is misleading

**Comment by fruch** [comment]
> > > LGTM if it works.
> > 
> > @soyacz let's have a more detailed plan on what needs to be covered.
> > 
> > I think we should try at least one type of testing, or wait for 6.0 branching before merging this.
> 
> I was testing it with provision test and analyzing logs if everything goes ok. I also executed Gemini test (as also oracle got parallelized). I was planning to run k8s multitenant case as it is not covered by provision tests triggered in by our SCT pipeline.
> 
> Anything else comes up to your mind?

Artifact tests, to see the minimal case works as expected 

And some LDAP logic was moved, maybe running one case for the safe side

File: sdcm/tester.py
Comment (soyacz) [reply sdcm/tester.py]: name fixed. thread moved out.

**Comment by soyacz** [comment]
ldap test failed, investigating (first checking if issue was not in master already)

**Comment by soyacz** [comment]
ms ldap test fails also without my change.
I did openldap test - passed.
Artifact tests also done (AMI based and Debian one) - look ok.

**Review by vponomaryov — APPROVED** [review]
LGTM

**Review by fruch — APPROVED** [review]
LGTM

**Pull request merged**
